sds: add file watcher support

ramaraochavali opened the pull request:
This PR 
- adds File Watcher support for file based SDS (Fixes #https://github.com/istio/istio/issues/23001)
- Also adds server certs in metadata to SDS config (https://github.com/istio/istio/issues/21833#issuecomment-615169432)

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 8400711e621a3b1b27246ba755aee7c3e7fd060f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/lint_istio/12225) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
@howardjohn Still I need to add tests for file watcher functionality - but rest of the changes are done. Want to get your early feedback. Can you PTAL?

howardjohn commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -247,27 +256,37 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
-		// Based on the resource name, we need to read this secret from a file encoded in the resource name
-		sdsFromFile = true
-		ns, err = sc.generateKeyCertFromExistingFiles(cfg.CertificatePath, cfg.PrivateKeyPath, token, connKey)
+		// It is sufficient if we add one of cert or key file for watching.
```
is there ever a case where one changes and not the other? is there harm in having both?

ramaraochavali replied:
I do not think so - but no harm in adding. I will add them

howardjohn commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +353,45 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+				select {
+				case <-timerC:
+					timerC = nil
+					if val, ok := sc.secrets.Load(key); ok {
```
What happens if the file we are watching is no longer present in any configs?

ramaraochavali replied:
I think we can remove it from watcher if Load fails here?

myidpt replied:
But the files are still being watched I think.
This code may have issues when used on the ingress gateway, which may watch many keys and certs.
Can you add a TODO here to safely remove the watchers for unused keys and certs?

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 8ae271684ae5be30e0ed58f2db55b0d3083e8639 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/lint_istio/12291) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 8ae271684ae5be30e0ed58f2db55b0d3083e8639 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/lint_istio/12291) | `/test lint_istio`
gencheck_istio | 8ae271684ae5be30e0ed58f2db55b0d3083e8639 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/gencheck_istio/10876) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 145d9d151d3473d9fbe70d6a3801f6a75596e9fb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/gencheck_istio/10877) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | bdaf25eab8d9d39f20b3d80a86e775751f5f2d12 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/gencheck_istio/10878) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | de089f1ab1e053d690a57eb9d680b9b005b7da3c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/unit-tests_istio/12795) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | de089f1ab1e053d690a57eb9d680b9b005b7da3c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/unit-tests_istio/12795) | `/test unit-tests_istio`
integ-mixer-k8s-tests_istio | de089f1ab1e053d690a57eb9d680b9b005b7da3c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-mixer-k8s-tests_istio/10704) | `/test integ-mixer-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-mixer-k8s-tests_istio | 3fb555fd2905528d70fc03945c972d59c6c6f1b6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-mixer-k8s-tests_istio/10706) | `/test integ-mixer-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-mixer-k8s-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-mixer-k8s-tests_istio | 3fb555fd2905528d70fc03945c972d59c6c6f1b6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-mixer-k8s-tests_istio/10707) | `/test integ-mixer-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-mixer-k8s-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-mixer-k8s-tests_istio | 3fb555fd2905528d70fc03945c972d59c6c6f1b6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-mixer-k8s-tests_istio/10708) | `/test integ-mixer-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-mixer-k8s-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-multicluster-k8s-tests_istio | 3fb555fd2905528d70fc03945c972d59c6c6f1b6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-multicluster-k8s-tests_istio/2797) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-multicluster-k8s-tests_istio

ramaraochavali commented:
@howardjohn @istio/wg-security-maintainers Added tests, this is ready for review

howardjohn commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +356,51 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+				case e := <-sc.certWatcher.Events(file):
+					if e.Op&fsnotify.Remove == fsnotify.Remove {
+						// File is deleted. Stop watching
+						cacheLog.Debugf("file deleted %s, removing the watcher", file)
```
if its later added back we should re add a watch. I believe this does occur as expected, just want to verify

ramaraochavali replied:
This is a good point. I removed the deletion logic because some systems delete and readd the new file. so it is safe not to remove from watch list. I also added a TODO to properly cleanup stale watches later.

lei-tang commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -187,6 +190,12 @@ type SecretCache struct {
 	existingKeyFile       string
 	existingRootCertFile  string
+
+	//certWatcher watches the certificates for changes and triggers a notification to proxy.
```
nit: "//certWatcher watches ..." should be "// certWatcher watches ..."

lei-tang commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -247,27 +258,38 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+		sc.addFileWatcher(sc.existingKeyFile, connKey)
+	default:
+		// Check if the resource name refers to a file mounted certificate.
+		// Currently Used in destination rules and server certs (via metadata).
```
nit: "Currently Used" should be "Currently used".

ramaraochavali commented:
@lei-tang Thanks for the review. I have addressed the comments and added a TODO for integration test which I will follow-up with after this is merged.

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +356,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 	return ns, nil
 }
 
+func (sc *SecretCache) addFileWatcher(file string, key ConnKey) {
```
Please use `connKey` instead of `key` for clarity. `key` can be misinterpreted as the private key.

ramaraochavali replied:
sounds good

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +356,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+					}
+					// Use a timer to debounce watch updates
+					if timerC == nil {
+						timerC = time.After(100 * time.Millisecond) // TODO: Make this configurable if needed.
```
So you introduced this to prevent two changing events of the private key and cert files to cause two secret push, correct? But there will still be two pushing events, with the same updated key and cert?

ramaraochavali replied:
Correct. There will be a push that sends both updated files because the push looks at the passed in secret item

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -247,27 +258,38 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+			sdsFromFile = true
+			ns, err = sc.generateKeyCertFromExistingFiles(cfg.CertificatePath, cfg.PrivateKeyPath, token, connKey)
+			sc.addFileWatcher(cfg.CertificatePath, connKey)
+			sc.addFileWatcher(cfg.PrivateKeyPath, connKey)
```
Can you just watch the cert not the key? A key can't change without changing the cert.
I'm asking because you may receive two update events for a key&cert pair update, and cause two push events. We don't want that to happen.

ramaraochavali replied:
Sure. That was my original implementation https://github.com/istio/istio/pull/23417#discussion_r418107281 - But added the other one later. I can go back to previous implementation

ramaraochavali commented:
@myidpt Thanks for your thorough review. Addressed all the feedback. PTAL

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | 044a04ff362e41b8e1e5d8ed318517a43668b4bb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-security-k8s-tests_istio/12562) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-local-tests_istio | e3af9bdd06ea5e7093bb71675398b382ba2fca85 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-local-tests_istio/3935) | `/test integ-local-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-local-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-local-tests_istio | e3af9bdd06ea5e7093bb71675398b382ba2fca85 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-local-tests_istio/3936) | `/test integ-local-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-local-tests_istio

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +323,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+					}
+				case e := <-sc.certWatcher.Events(file):
+					if len(e.Op.String()) > 0 { // To avoid spurious events, mainly coming from tests.
+						// Use a timer to debounce watch updates
```
Sorry, can you help me to understand how this debounce solves the spurious events issue? If every event is delated 100ms in processing, we still have a burst of events, just 100ms later, no?

ramaraochavali replied:
This debounce does not solve the spurious events case. Spurious events are mainly coming from tests injection. What this debounce solves is, let us say if key is modified, we will get an event, then later cert is modified we will get another event - This debounce helps to combine both events in to one and generate secret item once and push it on the SDS connection once.

myidpt replied:
Got it. You are nullifying timerC to do the debounce. Thanks!

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +323,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+						if _, secret, err := sc.generateFileSecret(connKey, token); err != nil {
+							cacheLog.Errorf("error in generating secret after file change %s, %v", file, err)
+						} else {
+							cacheLog.Debugf("file changed %s, triggering push to proxy", file)
```
Please make this Info level.

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +323,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+	sc.certMutex.Unlock()
+	// File is not being watched, start watching now and trigger key push.
+	if !exists {
+		cacheLog.Debugf("adding watcher for file %s", file)
```
Please make this Info level.

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +323,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+	// Check if this file is being already watched, if so ignore it. FileWatcher has the functionality of
+	// checking for duplicates. This check is needed here to avoid processing duplicate events for the same file.
+	sc.certMutex.Lock()
+	_, exists := sc.fileCerts[file]
```
1. Can you make sure this path is normalized so that different path names that resolving to the same file can be treated as the same one.
2. What about multiple connections are referring to the same file? I think this is possible for the gateway, or with customized destination rules. Add @JimmyCYJ to comment.

ramaraochavali replied:
1. Done
2. That is a great point. DRs will have definitely that case. I handled that. PTAL

ramaraochavali commented:
@myidpt No problem. Thank you for your review. Addressed comments. PTAL.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-local-tests_istio | 1c580c5967f7a577065d4603c0bdbdbcb9a625eb | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-local-tests_istio/3940) | `/test integ-local-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-local-tests_istio

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+	}
+	sc.certMutex.Unlock()
+	// File is not being watched, start watching now and trigger key push.
+	if !exists {
```
nit: early return if the file and connKey exist, to reduce the indention.

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+			sc.fileCerts[npath] = make(map[ConnKey]struct{})
+		}
+		sc.fileCerts[npath][connKey] = struct{}{}
+	}
```
Do you need to check and add the connKey to sc.fileCerts[npath] when the npath exists?

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+							if _, secret, err := sc.generateFileSecret(ckey, token); err != nil {
+								cacheLog.Errorf("error in generating secret after file change %s, %v", file, err)
+							} else {
+								cacheLog.Infof("file changed %s, triggering push to proxy", file)
```
Please add the connKey to the log.

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+						if _, ok := sc.secrets.Load(ckey); ok {
+							// Regenerate the Secret and trigger the callback that pushes the secrets to proxy.
+							if _, secret, err := sc.generateFileSecret(ckey, token); err != nil {
+								cacheLog.Errorf("error in generating secret after file change %s, %v", file, err)
```
Please add the connKey to the log.

ramaraochavali commented:
@myidpt Simplified the check as per your suggestion and also added connKey to logs

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 1c5a985ad2fd93c53ef9f65b8f2d9dd7829c349c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/gencheck_istio/10947) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test gencheck_istio

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+						if _, secret, err := sc.generateFileSecret(ckey, token); err != nil {
+							cacheLog.Errorf("error in generating secret after file change %s, %v, %v", file, err, ckey)
+						} else {
+							cacheLog.Infof("file changed %s, triggering push to proxy %v", file, ckey)
```
cacheLog.Infof("%v: file changed, triggering secret push to proxy [%s]", ckey, file)

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+					if _, ok := sc.secrets.Load(ckey); ok {
+						// Regenerate the Secret and trigger the callback that pushes the secrets to proxy.
+						if _, secret, err := sc.generateFileSecret(ckey, token); err != nil {
+							cacheLog.Errorf("error in generating secret after file change %s, %v, %v", file, err, ckey)
```
cacheLog.Errorf("%v: error in generating secret after file change [%s] %v", ckey, file, err)

myidpt commented on security/pkg/nodeagent/cache/secretcache.go:
```diff
@@ -335,6 +324,62 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
+	// File is not being watched, start watching now and trigger key push.
+	cacheLog.Infof("adding watcher for file %s", file)
+	if err := sc.certWatcher.Add(file); err != nil {
+		cacheLog.Errorf("error adding watcher for file %s, Skipping watches ", file)
```
cacheLog.Errorf("%v: error adding watcher for file, skipping watches [%s] %v", connKey, file, err)

ramaraochavali commented:
no problem. Changed. PTAL

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-local-tests_istio | 9d5945d26049b4bd25c30471d87d5dc164b95d19 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-local-tests_istio/3963) | `/test integ-local-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-local-tests_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-local-tests_istio | 9d5945d26049b4bd25c30471d87d5dc164b95d19 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23417/integ-local-tests_istio/4021) | `/test integ-local-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-local-tests_istio

rmichela commented:
@howardjohn Can this be backported to 1.6, or if possible, even 1.5.x?

rshriram commented:
/cherrypick release-1.6

istio-testing commented:
@rshriram: #23417 failed to apply on top of branch "release-1.6":
```Applying: add server certs
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M	pilot/pkg/model/authentication.go
M	security/pkg/nodeagent/cache/secretcache.go
Falling back to patching base and 3-way merge...
Auto-merging security/pkg/nodeagent/cache/secretcache.go
CONFLICT (content): Merge conflict in security/pkg/nodeagent/cache/secretcache.go
Auto-merging pilot/pkg/model/authentication.go
CONFLICT (content): Merge conflict in pilot/pkg/model/authentication.go
Patch failed at 0001 add server certs

```

<details>

In response to [this](https://github.com/istio/istio/pull/23417#issuecomment-624100867):

>/cherrypick release-1.6


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

rmichela commented:
I believe this PR needs to be cherry picked after #22019 has merged due to a dependency.

rmichela commented:
@rshriram @howardjohn Can you please reattempt this backport?

rshriram commented:
/cherrypick release-1.6

istio-testing commented:
@rshriram: new pull request created: #23636

<details>

In response to [this](https://github.com/istio/istio/pull/23417#issuecomment-625561881):

>/cherrypick release-1.6


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

rshriram commented:
FYI you can also manually create the PRs. The release managers would know to approve the proper ones automatically

rshriram commented:
/cherrypick release-1.6

istio-testing commented:
@rshriram: failed to push cherry-picked changes in GitHub: pushing failed, output: "To https://github.com/istio-testing/istio\n ! [rejected]              cherry-pick-23417-to-release-1.6 -> cherry-pick-23417-to-release-1.6 (non-fast-forward)\nerror: failed to push some refs to 'https://istio-testing:CENSORED@github.com/istio-testing/istio'\nhint: Updates were rejected because the tip of your current branch is behind\nhint: its remote counterpart. Integrate the remote changes (e.g.\nhint: 'git pull ...') before pushing again.\nhint: See the 'Note about fast-forwards' in 'git push --help' for details.\n", error: exit status 1

<details>

In response to [this](https://github.com/istio/istio/pull/23417#issuecomment-625562297):

>/cherrypick release-1.6


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
